File: Stiver_Problems/Arrays/P016.cpp
// P016.cpp - Maximum product subarray (brute + optimal)

#include <bits/stdc++.h>
using namespace std;

long long brute(const vector<int>& nums) {
    int n = nums.size();
    long long max_prod = LLONG_MIN;
    for (int i = 0; i < n - 1; i++) {
        long long prod = 1;
        for (int j = i; j < n; j++) {
            prod *= nums[j];
            max_prod = max(max_prod, prod);
        }
    }
    return max_prod;
}

long long optimal(const vector<int>& nums) {
    int n = nums.size();
    long long min_prod = nums[0], max_prod = nums[0], result = nums[0];
    for (int i = 1; i < n; i++) {
        long long cur = nums[i];
        if (cur < 0) swap(min_prod, max_prod);
        min_prod = min(cur, min_prod * cur);
        max_prod = max(cur, max_prod * cur);
        result = max(result, max_prod);
    }
    return result;
}

int main() {
    cout << "Enter nums: ";
    string line;
    getline(cin, line);
    stringstream ss(line);
    vector<int> nums;
    int x;
    while (ss >> x) nums.push_back(x);

    cout << "Brute: " << brute(nums) << "\n";
    cout << "Optimal: " << optimal(nums) << "\n";
    return 0;
}

File: Arrays/P016.py
def brute(nums):
    n=len(nums)
    max_prod=float('-inf')
    for i in range(n-1):
        prod=1
        for j in range(i,n):
            prod*=nums[i]
        max_prod=max(max_prod,prod)
    return max_prod

def optimal(nums):
    n=len(nums)
    min_prod=max_prod=result=nums[0]
    for i in range(1,n-1):
        if nums[i]<0:
            min_prod,max_prod=max_prod,min_prod
        min_prod=min(nums[i],min_prod*nums[i])
        max_prod=max(max_prod*nums[i],nums[i])
        result=max(result,max_prod)
    return result

if __name__=='__main__':
    nums=input('Enter nums: ').split()
    nums=list(map(int,nums))
    print('Brute: ',brute(nums))
    print('Optimal: ',optimal(nums))